src/bson.c

```diff
@@ -1201,6 +1289,230 @@ static bool php_phongo_bson_state_parse_type(zval* options, const char* name, ph
 	return retval;
 } /* }}} */
 
+static void field_path_map_element_set_info(php_phongo_field_path_map_element* element, php_phongo_bson_typemap_types type, zend_class_entry* ce)
+{
+	element->node_type = type;
+	element->node_ce   = ce;
+}
+
+static void map_add_field_path_element(php_phongo_bson_typemap* map, php_phongo_field_path_map_element* element)
+{
+	/* Make sure we have allocated enough */
+	if (map->field_paths.allocated_size < map->field_paths.size + 1) {
+		map->field_paths.allocated_size += PHONGO_FIELD_PATH_EXPANSION;
+		map->field_paths.map = erealloc(map->field_paths.map, sizeof(php_phongo_field_path_map_element) * map->field_paths.allocated_size);
+	}
+
+	map->field_paths.map[map->field_paths.size] = element;
+	map->field_paths.size++;
+}
+
+static php_phongo_field_path_map_element* field_path_map_element_alloc(void)
+{
+	php_phongo_field_path_map_element* tmp = ecalloc(1, sizeof(php_phongo_field_path_map_element));
+
+	tmp->entry = php_phongo_field_path_alloc(true);
+
+	return tmp;
+}
+
+static void field_path_map_element_dtor(php_phongo_field_path_map_element* element)
+{
+	php_phongo_field_path_free(element->entry);
+	efree(element);
+}
+
+bool php_phongo_bson_state_add_field_path(php_phongo_bson_typemap* map, char* field_path_original, php_phongo_bson_typemap_types type, zend_class_entry* ce TSRMLS_DC)
+{
+	char*                              ptr         = NULL;
+	char*                              segment_end = NULL;
+	php_phongo_field_path_map_element* field_path_map_element;
+
+	if (field_path_original[0] == '.') {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "A 'fieldPaths' key may not start with a '.'");
+		return false;
+	}
+
+	if (field_path_original[strlen(field_path_original) - 1] == '.') {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "A 'fieldPaths' key may not end with a '.'");
+		return false;
```

Comment:
Assuming both of these cases are handled by the "empty segment" error below, I'd propose allowing that to catch this. IMO, `".foo"`, `"foo..bar"`, and `"foo."` all contain empty segments.

Reply:
I had that first, but thought it better we have more precise error messages. This one is also faster to check as nothing has been allocated yet.